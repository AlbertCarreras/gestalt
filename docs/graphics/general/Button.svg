<svg width="325" height="180" viewBox="0 0 325 180" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="325" height="180" fill="#F6FDF5"/>
<path d="M119 90C119 76.7452 129.745 66 143 66H182C195.255 66 206 76.7452 206 90C206 103.255 195.255 114 182 114H143C129.745 114 119 103.255 119 90Z" fill="#005C62"/>
<path d="M141.069 94.8274C143.478 94.8274 144.972 93.6344 144.972 91.7285C144.972 90.3329 143.884 89.2599 142.39 89.1473V89.0123C143.548 88.8472 144.432 87.8492 144.432 86.7012C144.432 85.028 143.126 84 140.936 84H136V94.8274H141.069ZM138.362 85.6808H140.333C141.46 85.6808 142.109 86.206 142.109 87.0914C142.109 87.9918 141.413 88.502 140.138 88.502H138.362V85.6808ZM138.362 93.1392V89.9952H140.412C141.804 89.9952 142.563 90.5355 142.563 91.5484C142.563 92.5839 141.827 93.1392 140.466 93.1392H138.362Z" fill="white"/>
<path d="M154.687 86.5662H152.41V91.3383C152.41 92.4714 151.8 93.1692 150.666 93.1692C149.618 93.1692 149.078 92.5764 149.078 91.4059V86.5662H146.802V91.9461C146.802 93.867 147.96 95 149.821 95C151.128 95 151.933 94.4598 152.34 93.4993H152.481V94.8274H154.687V86.5662Z" fill="white"/>
<path d="M157.62 84.5928V86.6412H156.274V88.2995H157.62V92.5989C157.62 94.2046 158.449 94.8574 160.545 94.8574C160.983 94.8574 161.406 94.8124 161.687 94.7599V93.1467C161.468 93.1692 161.312 93.1842 161.014 93.1842C160.24 93.1842 159.896 92.839 159.896 92.0962V88.2995H161.687V86.6412H159.896V84.5928H157.62Z" fill="white"/>
<path d="M164.315 84.5928V86.6412H162.97V88.2995H164.315V92.5989C164.315 94.2046 165.144 94.8574 167.241 94.8574C167.679 94.8574 168.101 94.8124 168.383 94.7599V93.1467C168.164 93.1692 168.007 93.1842 167.71 93.1842C166.936 93.1842 166.591 92.839 166.591 92.0962V88.2995H168.383V86.6412H166.591V84.5928H164.315Z" fill="white"/>
<path d="M174.046 95C176.697 95 178.293 93.3943 178.293 90.693C178.293 88.0143 176.674 86.3861 174.046 86.3861C171.418 86.3861 169.798 88.0218 169.798 90.693C169.798 93.3868 171.394 95 174.046 95ZM174.046 93.2742C172.826 93.2742 172.137 92.3288 172.137 90.693C172.137 89.0723 172.833 88.1194 174.046 88.1194C175.25 88.1194 175.954 89.0723 175.954 90.693C175.954 92.3288 175.258 93.2742 174.046 93.2742Z" fill="white"/>
<path d="M180.155 94.8274H182.431V90.0477C182.431 88.9297 183.08 88.2094 184.136 88.2094C185.215 88.2094 185.724 88.8172 185.724 89.9802V94.8274H188V89.485C188 87.5116 186.968 86.3861 185.051 86.3861C183.776 86.3861 182.9 86.9638 182.501 87.8943H182.36V86.5662H180.155V94.8274Z" fill="white"/>
<path d="M0 0V-2H-2V0H0ZM325 0H327V-2H325V0ZM325 180V182H327V180H325ZM0 180H-2V182H0V180ZM0 2H325V-2H0V2ZM323 0V180H327V0H323ZM325 178H0V182H325V178ZM2 180V0H-2V180H2Z" fill="#767676"/>
</svg>
