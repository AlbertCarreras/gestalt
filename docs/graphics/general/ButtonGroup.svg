<svg width="325" height="180" viewBox="0 0 325 180" fill="none" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_102_2537)">
<rect width="325" height="180" fill="#F6FDF5"/>
<path d="M0 0V-1H-1V0H0ZM325 0H326V-1H325V0ZM325 180V181H326V180H325ZM0 180H-1V181H0V180ZM0 1H325V-1H0V1ZM324 0V180H326V0H324ZM325 179H0V181H325V179ZM1 180V0H-1V180H1Z" fill="#767676"/>
<path d="M166 90C166 76.7452 176.745 66 190 66H229C242.255 66 253 76.7452 253 90C253 103.255 242.255 114 229 114H190C176.745 114 166 103.255 166 90Z" fill="#005C62"/>
<path d="M188.069 94.8274C190.478 94.8274 191.972 93.6344 191.972 91.7285C191.972 90.3329 190.884 89.2599 189.39 89.1473V89.0123C190.548 88.8472 191.432 87.8492 191.432 86.7012C191.432 85.028 190.126 84 187.936 84H183V94.8274H188.069ZM185.362 85.6808H187.333C188.46 85.6808 189.109 86.206 189.109 87.0914C189.109 87.9918 188.413 88.502 187.138 88.502H185.362V85.6808ZM185.362 93.1392V89.9952H187.412C188.804 89.9952 189.563 90.5355 189.563 91.5484C189.563 92.5839 188.827 93.1392 187.466 93.1392H185.362Z" fill="white"/>
<path d="M201.687 86.5662H199.41V91.3383C199.41 92.4714 198.8 93.1692 197.666 93.1692C196.618 93.1692 196.078 92.5764 196.078 91.4059V86.5662H193.802V91.9461C193.802 93.867 194.96 95 196.821 95C198.128 95 198.933 94.4598 199.34 93.4993H199.481V94.8274H201.687V86.5662Z" fill="white"/>
<path d="M204.62 84.5928V86.6412H203.274V88.2995H204.62V92.5989C204.62 94.2046 205.449 94.8574 207.545 94.8574C207.983 94.8574 208.406 94.8124 208.687 94.7599V93.1467C208.468 93.1692 208.312 93.1842 208.014 93.1842C207.24 93.1842 206.896 92.839 206.896 92.0962V88.2995H208.687V86.6412H206.896V84.5928H204.62Z" fill="white"/>
<path d="M211.315 84.5928V86.6412H209.97V88.2995H211.315V92.5989C211.315 94.2046 212.144 94.8574 214.241 94.8574C214.679 94.8574 215.101 94.8124 215.383 94.7599V93.1467C215.164 93.1692 215.007 93.1842 214.71 93.1842C213.936 93.1842 213.591 92.839 213.591 92.0962V88.2995H215.383V86.6412H213.591V84.5928H211.315Z" fill="white"/>
<path d="M221.046 95C223.697 95 225.293 93.3943 225.293 90.693C225.293 88.0143 223.674 86.3861 221.046 86.3861C218.418 86.3861 216.798 88.0218 216.798 90.693C216.798 93.3868 218.394 95 221.046 95ZM221.046 93.2742C219.826 93.2742 219.137 92.3288 219.137 90.693C219.137 89.0723 219.833 88.1194 221.046 88.1194C222.25 88.1194 222.954 89.0723 222.954 90.693C222.954 92.3288 222.258 93.2742 221.046 93.2742Z" fill="white"/>
<path d="M227.155 94.8274H229.431V90.0477C229.431 88.9297 230.08 88.2094 231.136 88.2094C232.215 88.2094 232.724 88.8172 232.724 89.9802V94.8274H235V89.485C235 87.5116 233.968 86.3861 232.051 86.3861C230.776 86.3861 229.9 86.9638 229.501 87.8943H229.36V86.5662H227.155V94.8274Z" fill="white"/>
<path d="M71 90C71 76.7452 81.7452 66 95 66H134C147.255 66 158 76.7452 158 90C158 103.255 147.255 114 134 114H95C81.7452 114 71 103.255 71 90Z" fill="#009990"/>
<path d="M93.0686 94.8274C95.4777 94.8274 96.9717 93.6344 96.9717 91.7285C96.9717 90.3329 95.8845 89.2599 94.3905 89.1473V89.0123C95.5481 88.8472 96.432 87.8492 96.432 86.7012C96.432 85.028 95.1258 84 92.9356 84H88V94.8274H93.0686ZM90.3622 85.6808H92.3333C93.4597 85.6808 94.1089 86.206 94.1089 87.0914C94.1089 87.9918 93.4128 88.502 92.1378 88.502H90.3622V85.6808ZM90.3622 93.1392V89.9952H92.4116C93.8039 89.9952 94.5626 90.5355 94.5626 91.5484C94.5626 92.5839 93.8273 93.1392 92.4663 93.1392H90.3622Z" fill="white"/>
<path d="M106.687 86.5662H104.41V91.3383C104.41 92.4714 103.8 93.1692 102.666 93.1692C101.618 93.1692 101.078 92.5764 101.078 91.4059V86.5662H98.802V91.9461C98.802 93.867 99.9597 95 101.821 95C103.128 95 103.933 94.4598 104.34 93.4993H104.481V94.8274H106.687V86.5662Z" fill="white"/>
<path d="M109.62 84.5928V86.6412H108.274V88.2995H109.62V92.5989C109.62 94.2046 110.449 94.8574 112.545 94.8574C112.983 94.8574 113.406 94.8124 113.687 94.7599V93.1467C113.468 93.1692 113.312 93.1842 113.014 93.1842C112.24 93.1842 111.896 92.839 111.896 92.0962V88.2995H113.687V86.6412H111.896V84.5928H109.62Z" fill="white"/>
<path d="M116.315 84.5928V86.6412H114.97V88.2995H116.315V92.5989C116.315 94.2046 117.144 94.8574 119.241 94.8574C119.679 94.8574 120.101 94.8124 120.383 94.7599V93.1467C120.164 93.1692 120.007 93.1842 119.71 93.1842C118.936 93.1842 118.591 92.839 118.591 92.0962V88.2995H120.383V86.6412H118.591V84.5928H116.315Z" fill="white"/>
<path d="M126.046 95C128.697 95 130.293 93.3943 130.293 90.693C130.293 88.0143 128.674 86.3861 126.046 86.3861C123.418 86.3861 121.798 88.0218 121.798 90.693C121.798 93.3868 123.394 95 126.046 95ZM126.046 93.2742C124.826 93.2742 124.137 92.3288 124.137 90.693C124.137 89.0723 124.833 88.1194 126.046 88.1194C127.25 88.1194 127.954 89.0723 127.954 90.693C127.954 92.3288 127.258 93.2742 126.046 93.2742Z" fill="white"/>
<path d="M132.155 94.8274H134.431V90.0477C134.431 88.9297 135.08 88.2094 136.136 88.2094C137.215 88.2094 137.724 88.8172 137.724 89.9802V94.8274H140V89.485C140 87.5116 138.968 86.3861 137.051 86.3861C135.776 86.3861 134.9 86.9638 134.501 87.8943H134.36V86.5662H132.155V94.8274Z" fill="white"/>
</g>
<path d="M0 0V-2H-2V0H0ZM325 0H327V-2H325V0ZM325 180V182H327V180H325ZM0 180H-2V182H0V180ZM0 2H325V-2H0V2ZM323 0V180H327V0H323ZM325 178H0V182H325V178ZM2 180V0H-2V180H2Z" fill="#767676"/>
<defs>
<clipPath id="clip0_102_2537">
<path d="M0 0H325V180H0V0Z" fill="white"/>
</clipPath>
</defs>
</svg>
