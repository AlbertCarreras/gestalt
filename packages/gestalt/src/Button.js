// @flow strict
import { type AbstractComponent, forwardRef, type Node } from 'react';
import InternalButton from './Button/InternalButton.js';
import icons from './icons/index.js';

type Props = {|
  /**
   * Text to render inside the Button to convey the function and purpose of the Button.
   */
  text: string,
  /**
   * A unique id indicating the element or elements whose contents or visibility are controlled by Button.
   * See the [Accessibility guidelines](https://gestalt.pinterest.systems/foundations/accessibility) to learn more.
   */
  accessibilityControls?: string,
  /**
   * Needed if Button controls the visibility of other elements, e.g. Dropdown or Flyout.
   * This is used to indicate if the controlled grouping is currently expanded or collapsed.
   * See the [Accessibility guidelines](https://gestalt.pinterest.systems/foundations/accessibility) to learn more.
   */
  accessibilityExpanded?: boolean,
  /**
   * Set as true if Button controls one or more interactive popup elements, such as a menu or dialog.
   * See the [Accessibility guidelines](https://gestalt.pinterest.systems/foundations/accessibility) to learn more.
   */
  accessibilityHaspopup?: boolean,
  /**
   * Label to provide more context around Buttonâ€™s function or purpose.
   * See the [Accessibility guidelines](https://gestalt.pinterest.systems/foundations/accessibility) to learn more.
   */
  accessibilityLabel?: string,
  /**
   * The background color of Button.
   */
  color?:
    | 'gray'
    | 'red'
    | 'blue'
    | 'transparent'
    | 'semiTransparentWhite'
    | 'transparentWhiteText'
    | 'white',
  /**
   * Available for testing purposes, if needed. Consider [better queries](https://testing-library.com/docs/queries/about/#priority) before using this prop.
   */
  dataTestId?: string,
  /**
   * Indicates if Button is disabled. Disabled Buttons are inactive and cannot be interacted with.
   */
  disabled?: boolean,
  /**
   * Default Buttons are sized by the text within the Button whereas full-width Buttons expand to the full width of their container.
   */
  fullWidth?: boolean,
  /**
   * An icon displayed after the text to help clarify the usage of Button. See the [icon variant](https://gestalt.pinterest.systems/web/button#Icons) to learn more.
   */
  iconEnd?: $Keys<typeof icons>,
  /**
   * The name attribute specifies the name of the \<button\> element.
   * The name attribute is used to reference form-data after the form has been submitted.
   */
  name?: string,
  /**
   * Callback invoked when the user clicks (press and release) on Button with the mouse or keyboard. Required with `role="button"` or `type="button"` Buttons.
   * See [GlobalEventsHandlerProvider](/web/utilities/globaleventshandlerprovider#Link-handlers) to learn more about link navigation.,
   */
  onClick?: ({|
    event: SyntheticMouseEvent<HTMLButtonElement> | SyntheticKeyboardEvent<HTMLButtonElement>,
    dangerouslyDisableOnNavigation: () => void,
  |}) => void,
  /**
   * Indicates if Button is currently selected.
   */
  selected?: boolean,
  /**
   * sm: 32px, md: 40px, lg: 48px
   */
  size?: 'sm' | 'md' | 'lg',
  /**
   * Use "-1" to remove Button from keyboard navigation.
   * See the [Accessibility guidelines](https://gestalt.pinterest.systems/foundations/accessibility) to learn more.
   */
  tabIndex?: -1 | 0,
  /**
   * sm: 32px, md: 40px, lg: 48px
   */
  type?: 'button' | 'submit',
|};

/**
 * [Buttons](https://gestalt.pinterest.systems/web/button) allow users to perform actions within a surface. They can be used alone for immediate action, or as a trigger for another component, like [Dropdown](https://gestalt.pinterest.systems/web/dropdown) or [Popover](https://gestalt.pinterest.systems/web/popover).
 *
 * ![Button light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Button.spec.mjs-snapshots/Button-chromium-darwin.png)
 * ![Button dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Button-dark.spec.mjs-snapshots/Button-dark-chromium-darwin.png)
 *
 */
const ButtonWithForwardRef: AbstractComponent<Props, HTMLButtonElement> = forwardRef<
  Props,
  HTMLButtonElement,
>(function Button(
  {
    text,
    accessibilityControls,
    accessibilityExpanded,
    accessibilityHaspopup,
    accessibilityLabel,
    color = 'gray',
    dataTestId,
    disabled = false,
    fullWidth = false,
    iconEnd,
    name,
    onClick,
    selected = false,
    size = 'md',
    tabIndex = 0,
    type,
  }: Props,
  ref,
): null | Node {
  if (type === 'submit') {
    return (
      <InternalButton
        accessibilityLabel={accessibilityLabel}
        dataTestId={dataTestId}
        disabled={disabled}
        fullWidth={fullWidth}
        name={name}
        // $FlowExpectedError[incompatible-type]
        ref={ref}
        role="button"
        type="submit"
        tabIndex={tabIndex}
        size={size}
        text={text}
        color={color}
        iconEnd={iconEnd}
        onClick={onClick}
      />
    );
  }
  return (
    <InternalButton
      accessibilityControls={accessibilityControls}
      accessibilityExpanded={accessibilityExpanded}
      accessibilityHaspopup={accessibilityHaspopup}
      accessibilityLabel={accessibilityLabel}
      dataTestId={dataTestId}
      disabled={disabled}
      fullWidth={fullWidth}
      name={name}
      // $FlowExpectedError[incompatible-type]
      ref={ref}
      role="button"
      type="button"
      tabIndex={tabIndex}
      selected={selected}
      size={size}
      text={text}
      color={color}
      iconEnd={iconEnd}
      onClick={onClick}
    />
  );
});

ButtonWithForwardRef.displayName = 'Button';

export default ButtonWithForwardRef;
