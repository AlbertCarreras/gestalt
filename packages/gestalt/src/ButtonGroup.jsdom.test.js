// @flow strict

import { render } from '@testing-library/react';
import Button from './Button.js';
import ButtonGroup from './ButtonGroup.js';

describe('ButtonGroup', () => {
  test('Renders nothing when no children are passed in', () => {
    const { container } = render(<ButtonGroup />);
    expect(container).toMatchSnapshot();
  });

  test('Renders the child without container when 1 child is passed in', () => {
    const { container } = render(
      <ButtonGroup>
        <Button text="Button 1" />
      </ButtonGroup>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Renders container with children when multiple children are passed in', () => {
    const { container } = render(
      <ButtonGroup>
        <Button text="Button 1" />
        <Button text="Button 2" />
      </ButtonGroup>,
    );
    expect(container).toMatchSnapshot();
  });
});
