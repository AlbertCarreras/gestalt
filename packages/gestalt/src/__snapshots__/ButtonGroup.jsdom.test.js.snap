// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup Renders container with children when multiple children are passed in 1`] = `
<div>
  <div
    class="box flexWrap marginBottomN1 marginEndN1 marginStartN1 marginTopN1 xsDisplayFlex"
  >
    <div
      class="box paddingX1 paddingY1"
    >
      <button
        class="button hideOutline inline accessibilityOutline parentButton"
        tabindex="0"
        type="button"
      >
        <div
          class="button hideOutline inline accessibilityOutline tapTransition md gray enabled childrenDiv"
        >
          <div
            class="Text fontSize300 defaultText alignCenter fontWeightSemiBold"
          >
            Button 1
          </div>
        </div>
      </button>
    </div>
    <div
      class="box paddingX1 paddingY1"
    >
      <button
        class="button hideOutline inline accessibilityOutline parentButton"
        tabindex="0"
        type="button"
      >
        <div
          class="button hideOutline inline accessibilityOutline tapTransition md gray enabled childrenDiv"
        >
          <div
            class="Text fontSize300 defaultText alignCenter fontWeightSemiBold"
          >
            Button 2
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`ButtonGroup Renders nothing when no children are passed in 1`] = `<div />`;

exports[`ButtonGroup Renders the child without container when 1 child is passed in 1`] = `
<div>
  <div
    class="box flexWrap marginBottomN1 marginEndN1 marginStartN1 marginTopN1 xsDisplayFlex"
  >
    <div
      class="box paddingX1 paddingY1"
    >
      <button
        class="button hideOutline inline accessibilityOutline parentButton"
        tabindex="0"
        type="button"
      >
        <div
          class="button hideOutline inline accessibilityOutline tapTransition md gray enabled childrenDiv"
        >
          <div
            class="Text fontSize300 defaultText alignCenter fontWeightSemiBold"
          >
            Button 1
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;
